servicetalk-http-api/src/main/java/io/servicetalk/http/api/OffloadingFilter.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.api;
+
+import io.servicetalk.concurrent.api.Executor;
+import io.servicetalk.concurrent.api.Single;
+
+import static java.util.function.Function.identity;
+
+/**
+ * An {@link StreamingHttpServiceFilterFactory} implementation which offloads filters using a provided strategy.
+ */
+final class OffloadingFilter implements StreamingHttpServiceFilterFactory {
+
+    private final HttpExecutionStrategy strategy;
+    private final StreamingHttpServiceFilterFactory offloaded;
+
+    /**
+     * @param strategy Execution strategy for the offloaded filters
+     * @param offloaded Filters to be offloaded
+     */
+    OffloadingFilter(HttpExecutionStrategy strategy, StreamingHttpServiceFilterFactory offloaded) {
+        this.strategy = strategy;
+        this.offloaded = offloaded;
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(StreamingHttpService service) {
+        return new StreamingHttpServiceFilter(offloaded.create(service)) {
+
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                Executor e = null != strategy.executor() ? strategy.executor() : ctx.executionContext().executor();
+
+                // The service should see our ExecutionStrategy inside the ExecutionContext:
+                final HttpServiceContext wrappedCtx =
```

Comment:
if `ExecutionStrategy` interface is reduced to just the state (e.g. boolean methods), will we need `ExecutionContext executionStrategy()` on the `ExecutionContext` interface (and by extension `HttpExecutionContext`)? perhaps this is still necessary for routers to see which strategy has been used thus far and do a diff?

Reply:
Yes, the diff is why it is still needed. I believe there are cases where this will be needed.

Reply:
IIUC after `Executor` is removed wrapping is only necessary if `!ctx.executionContext().executionStrategy().equals(strategy)`? I expect this `equals` comparison should be cheap after this point (coming like `int` bit mask comparison).

Reply:
Correct. The comparison between execution strategies is already fairly cheap anyway due to short-circuiting.